src/number_theory/modular_forms/slash_invariant_forms.lean

```diff
@@ -0,0 +1,209 @@
+
+@[priority 100]
+instance slash_invariant_form_class.slash_invariant_form :
+   slash_invariant_form_class (slash_invariant_form Γ k) Γ k :=
+{ coe := (slash_invariant_form.to_fun),
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := slash_invariant_form.slash_action_eq' }
+
+variables {F Γ k}
+
+instance : has_coe_to_fun (slash_invariant_form Γ k) (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+
+@[simp] lemma slash_invariant_form_to_fun_eq_coe {f : slash_invariant_form Γ k} :
+  f.to_fun = (f : ℍ → ℂ) := rfl
+
+@[ext] theorem slash_invaraint_form_ext {f g : slash_invariant_form Γ k} (h : ∀ x, f x = g x) :
+  f = g := fun_like.ext f g h
+
+/-- Copy of a `slash_invariant_form` with a new `to_fun` equal to the old one.
+Useful to fix definitional equalities. -/
+protected def slash_invariant_form.copy (f : slash_invariant_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  slash_invariant_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',}
+
+end slash_invariant_forms
+
+namespace slash_invariant_form
+
+open slash_invariant_form
+
+variables {F : Type*} {Γ : out_param $ subgroup SL(2, ℤ)} {k : out_param ℤ}
+
+@[priority 100, nolint dangerous_instance]
+instance slash_invariant_form_class.coe_to_fun [slash_invariant_form_class F Γ k] :
+  has_coe_to_fun F (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+
+@[simp] lemma slash_action_eqn [slash_invariant_form_class F Γ k] (f : F) (γ : Γ) :
+   slash_action.map ℂ k γ ⇑f = ⇑f := slash_invariant_form_class.slash_action_eq f γ
+
+lemma slash_action_eqn' (k : ℤ) (Γ : subgroup SL(2, ℤ)) [slash_invariant_form_class F Γ k] (f : F) :
+∀ γ : Γ, ∀ z : ℍ, f (γ • z) = ((↑ₘγ 1 0 : ℝ) * z +(↑ₘγ 1 1 : ℝ))^k * f z :=
+begin
+  intros γ z,
+  rw ←modular_form.slash_action_eq'_iff,
+  simp,
+end
+
+instance [slash_invariant_form_class F Γ k] : has_coe_t F (slash_invariant_form Γ k) :=
+⟨λ f, { to_fun := f, slash_action_eq' := slash_action_eqn f }⟩
+
+@[simp] lemma slash_invariant_form_class.coe_coe [slash_invariant_form_class F Γ k] (f : F) :
+  ((f : (slash_invariant_form Γ k)) : ℍ → ℂ) = f := rfl
+
+instance has_add : has_add (slash_invariant_form Γ k) :=
+⟨λ f g , ⟨ f + g, by {intro γ, convert slash_action.add_action k γ (f : ℍ → ℂ) g,
+   exact ((f.slash_action_eq') γ).symm, exact ((g.slash_action_eq') γ).symm} ⟩⟩
+
+@[simp] lemma coe_add (f g : slash_invariant_form Γ k) : ⇑(f + g) = f + g := rfl
+@[simp] lemma add_apply (f g : slash_invariant_form Γ k) (z : ℍ) : (f + g) z = f z + g z := rfl
+
+instance has_zero : has_zero (slash_invariant_form Γ k) :=
+{zero := ⟨ 0 , slash_action.mul_zero _⟩}
+
+lemma nsmul_coe [slash_invariant_form_class F Γ k] (f : F) (c : ℕ) :
+  c • (f : ℍ → ℂ) = (c : ℂ) • f :=
+begin
+ simp only [nsmul_eq_mul],
+ congr,
+end
+
+lemma zsmul_coe [slash_invariant_form_class F Γ k] (f : F) (c : ℤ) :
+  c • (f : ℍ → ℂ) = (c : ℂ) • f :=
+begin
+ simp only [zsmul_eq_mul],
+ congr,
+end
+
+instance has_nsmul : has_smul ℕ (slash_invariant_form Γ k) :=
```

Comment:
I propose moving this to after the `csmul` definition and then simply defining it as 
```lean
instance has_nsmul : has_smul ℕ (slash_invariant_form Γ k) :=
⟨ λ c f, ((c:ℂ) • f).copy (c • f) (nsmul_eq_smul_cast _ _ _) ⟩
```
This gets rid of the proof here and of the lemma `nsmul_coe`.  Similar for `zsmul`.

Reply:
I've kept `nsmul_coe` since its needed for the `add_comm_group` instance.

Reply:
You can replace it at the point of use by `nsmul_eq_smul_cast` (it's a special case of that lemma) and delete it.

Reply:
ok sure :)

Reply:
Sorry, maybe I wasn't clear.  I'm proposing getting rid of the lemmas you had called `nsmul_coe` and `zsmul_coe`.  The lemmas you had called `coe_nsmul` and `coe_zsmul` are fine in my opinion!

Reply:
Oh sorry I misread the names! I've fixed this now :)